### ISSUE TITLE
This app is blocked by Google

### ISSUE THREAD

**Issue by jet76** [Kyrodan/KeeAnywhere#438]
**Describe the bug**
This app is blocked
This app tried to access sensitive info in your Google Account. To keep your account safe, Google blocked this access.

**To Reproduce**
1. Authenticate Google Drive

**Expected behavior**
Authentication succeeds

**Environment:**
 - OS and Version: Windows 11
 - KeePass-Version: 2.57.1
 - KeeAnywhere-Version:  2.1.0

**Comment by bknickelbine** [comment]
@jet76 Have you tried removing the third-party app connection in your Google Account, then setting up a GoogleDriveRestricted account in KeeAnywhere? That seemed to work for me, at least while the main issue is being worked on.

**Comment by Karreg** [comment]
Same issue here.
I tried with the GoogleDriveRestricted account, and it's working indeed, but I can't browser my drive content to open the password database.
I was not able to find documentation on this rescripted account, should we give the path manually without browsing?

**Comment by AGheller** [comment]
I also encountered the same problem. I switched to using GoogleDriveRestricted. I then downloaded the database in Google Drive from the browser, then opened it with KeepPass and had it saved again from there. With GoogleDriveRestricted you only have access to the files created by the application, so it is normal to no longer see everything else.

**Comment by Karreg** [comment]
I don't understand what you did... What is the difference between downloading manually to the drive folder, and synchronizing in the drive folder from the drive app? Also, if the file exists, it was not created from the app so it can't be seen either? Also, how does it work with multiple devices? This workaround does only work for a single device setup right?

**Comment by Karreg** [comment]
In fact the issue already exists in #436 with mode information 👍

**Comment by splisp** [comment]
Same here since this morning

![image](https://github.com/user-attachments/assets/6504fc0b-237d-4dea-bca4-9602876b822a)

**Comment by JuergenHe** [comment]
I have got the same problem. It worked fine until yesterday.
Yes - using restricted is possible. But this is a very cumbersome workaround and not really useful.

**Comment by AGheller** [comment]
> I don't understand what you did... What is the difference between downloading manually to the drive folder, and synchronizing in the drive folder from the drive app? Also, if the file exists, it was not created from the app so it can't be seen either? Also, how does it work with multiple devices? This workaround does only work for a single device setup right?

I mean I save the file again from KeePass, the application is KeyAnywhere and it is the same on every devices.

**Comment by ItalianInPoland** [comment]
> > I don't understand what you did... What is the difference between downloading manually to the drive folder, and synchronizing in the drive folder from the drive app? Also, if the file exists, it was not created from the app so it can't be seen either? Also, how does it work with multiple devices? This workaround does only work for a single device setup right?
> 
> I mean I save the file again from KeePass, the application is KeyAnywhere and it is the same on every devices.

Hi, I confirm it works.

**Comment by jonahhes** [comment]
> > I don't understand what you did... What is the difference between downloading manually to the drive folder, and synchronizing in the drive folder from the drive app? Also, if the file exists, it was not created from the app so it can't be seen either? Also, how does it work with multiple devices? This workaround does only work for a single device setup right?
> 
> I mean I save the file again from KeePass, the application is KeyAnywhere and it is the same on every devices.

This won't work for interoperability with other platforms like Keepass2Android though, right?

**Comment by ItalianInPoland** [comment]
it always works for me. With keeanywhere for PC and Keepass2Android for phones (both owner of the file on gdrive and shared).

**Comment by GreyMatter-TRTH** [comment]
Is it time to say goodbye..😟? Nothing seems to work...🤔

**Comment by hamidamadani** [comment]
Same problem. It was working as expected till yesterday.

**Comment by jonahhes** [comment]
@meokey Actually I tied it and for me it does work as @ItalianInPoland described with GDrive restricted.

Steps:
1. Add GDrive restricted in KeeAnywhere settings (I used the same google acc't as normal GDrive)
2. Save as... -> Save to cloud drive -> GDrive restricted -> The folder you see is empty, just save it there 
3. You should now see the new file in your GDrive, in the root folder, besides all the other stuff that was there but KeeAnywhere couldn't see
4. In Keepass2Android select Open file -> Google Drive
5. You should be able to select the new file because Keepass2Android does not run in restricted mode (yet?)
6. On Keepass Desktop, you might also have to reopen the file from GDrive restricted.

That's it. Works for now and syncs both ways.

**Comment by Sepp-PS** [comment]
Since today I have the same problem. I tried first a new reconnect with GDrive, after that with GDrive restricted. But it will not work.
Error message with GDrive restricted:
Google.Apis.Requests.RequestError
Request had insufficient Authentication Scopes. [403]
....

I'm sure that I made the connection correctly. Then I found here the open issue and report my statues as well. 
Hopefully you can fix the changes from Google with and adjustment of KeeAnywhere.
Keepass 2.57.1
KeeAnywhere 2.1.0
Windows 11 23H2

**Comment by bigboj** [comment]
gdrive:// was working fine until yesterday. All of sudden, one of my PCs started complaining about auth issue and using the cached file, but another one of my PCs was still working, so I went through https://github.com/Kyrodan/KeeAnywhere/issues/393 and unprotect/reprotect'd my key from the working OS, which seemed to fix the issue.

Today, both all of my PCs failed to open the gdrive:// DBs. 

Checking KeeAnywhere permissions in Google Account's [third-party app list](https://myaccount.google.com/connections?continue=https%3A%2F%2Fmyaccount.google.com%2Fdata-and-privacy), I can see it only has `See, edit, create and delete only the specific Google Drive files that you use with this app`.
While I cannot recall exactly, but I'm almost sure, yesterday there was `See, edit, create and delete all of your Google Drive files` permissions in addition.

Eventually, I accepted the only option is to save my file as gdrive-r:// (Google Drive restricted), however it comes with two issues:
1., (less serious) - it can only save to root directory. I'm just not comfortable with this.
2., (much more serious - may affect many users) - Sharing the file with other google users does not work. (as far as I was able to try all I could). This issue doesn't seem to have any solution or even workaround. 

I also use KeePass2Android, which is still working fine (yet?). I can open any of my kdbx files either directly, using "Google Drive" or "Google Drive (KP2A files)" storage types, or by using the system file picker and navigate to files hosted by my local Google Drive app installed. (Also noted, KP2A still has the `See, edit, create and delete all of your Google Drive files` permission)

**Comment by hamidamadani** [comment]
According to another topic on the net, I just checked "Use limited drive access" in the configuration dialog of "KPsync" and the problem resolved.

**Comment by hamidamadani** [comment]
I use KeePass for Windows with the KPsync plugin. In the `Configure` section of the plugin, be sure to checkmark the `Use limited drive access option`. The file is in `restricted sharing` mode in my Google drive.
@Sepp-PS

**Comment by Sepp-PS** [comment]
An Update from my tests. 
I don't need anymore the "normal" feature. The restricted version work as well. Today new files where I have created with KeeAnywhere shown me latest after 1 minute in the local GDrive Sync folder.
For me: Problem solved.

**Comment by f1h0** [comment]
Colleagues, what should I do to make it work? inside KeeAnywhere I click check account verification and then go back to the Google application and then the information that was blocked. what should I do, please explain ?

**Comment by lammertsm** [comment]
Same issue

**Comment by Sepp-PS** [comment]
In the case where you use GDrive in the old way you can't change it to the needed new setup. (So far what I see).
You must open the current KDBX file from GDrive (client version) and must save it in a new entry. In that new entry you must use GDrive restricted. Select you account in the browser and in the next page you must select the check mark for GDrive. 
That's it. 

Or with other words: A verification from the old setup / GDrive normal can't work. Make it new (not the KDBX file)

**Comment by f1h0** [comment]
а в связи с чем это такое произошло  раньше всё работало

> In the case where you use GDrive in the old way you can't change it to the needed new setup. (So far what I see). You must open the current KDBX file from GDrive (client version) and must save it in a new entry. In that new entry you must use GDrive restricted. Select you account in the browser and in the next page you must select the check mark for GDrive. That's it.
> 
> Or with other words: A verification from the old setup / GDrive normal can't work. Make it new (not the KDBX file)
What caused this to happen? It was working before.

**Comment by Sepp-PS** [comment]
I assume Google increased the security and will phaseout old API's. I'm not a developer in that area, but I'm happy that the developer of KeeAnywhere  has that feature.

**Comment by mhobl** [comment]
I do have the same issue.

This is what I have found officially from Google:
https://support.google.com/accounts/answer/6010255?hl=en

It seems they have already started to roll-out some of those changes.

**Comment by Karreg** [comment]
> I do have the same issue.
> 
> This is what I have found officially from Google: https://support.google.com/accounts/answer/6010255?hl=en
> 
> It seems they have already started to roll-out some of those changes.

To me this is not the issue. The message is different, and KeeAnywhere does not use login/password. This is something that has been obsolete for years, and is not compatible avec 2FA anyway.

Please go to #436 as this is a duplicate

**Comment by alexmuzenhardt** [comment]
I just switched to Dropbox until this issue is solved for Gdrive. Works like a charm. 🙂

**Comment by nascentt** [comment]
This has been broken for me for a week too.  I ended up switching to GDrive restricted as above, which allows keeanywhere to save and load from it, but no other apps can access the database so it's fairly useless as a sync target.

**Comment by Sepp-PS** [comment]
> This has been broken for me for a week too. I ended up switching to GDrive restricted as above, which allows keeanywhere to save and load from it, but no other apps can access the database so it's fairly useless as a sync target.

I can't agree with that statement. I use from more devices via the restricted mode the sync option. And it is more secure - only KeeAnywhere should access the data. And you can always use the native GDrive Client App to what every you want. So other app where don't must use the restricted way have access. The question is only which one will not block Google?

**Comment by Sepp-PS** [comment]
If the app has full access to GDrive, it has also access to the file. Only the app, where have restricted access, can see the own created files. The same app, on on 2nd device can see it as well - via the restricted mode.

**Comment by nascentt** [comment]
As Sepp-PS says, I cleared the old keepass dB from my android app (keepass2android) and managed to search and find the file that keepass anywhere had saved to GDrive (Restricted). So things are now syncing again.

**Comment by CrashO86** [comment]
Hello, is there any news regarding this bug?

**Comment by Sepp-PS** [comment]
@CrashO86 : Do you thinking that Google revoke the app blocking? I don't thinking that will be happen. 
And it is no bug -  more a solved issuer. Please switch to GDrive Restricted.

**Comment by f1h0** [comment]
google megamonster dictates its rules you have to put up with it or find some other alternative solutions

**Comment by Kyrodan** [comment]
Hi together,

I read all your comments and be myself unhappy about this situation.
@f1h0 wrote;
> google megamonster dictates its rules you have to put up with it or find some other alternative solutions
From my perspective this is a very good summary.

I would like to explain my standpoint:
* Google has rules to follow without exception. They announce changes early so developers have enough time to adapt.
* Many of these rule changes do make sense from user's perspective. But some decision are very "uncommon".

In this case Google decided to change a very strategic topic:
In the past each developer was able to request a "unrestricted" scope for GDrive and the user can accept this or not in the consent dialog. This was okay for nearly all users: they know what they do. The other option for developers is to restrict the apps to only use "it's own files". This in general okay, too. The same thing has been implemented some time ago for Dropbox. But in contrast to Dropbox the situation is a little bit different: For unrestricted access Google now requests developers to apply for an app review. Developers have to send in several artifacts (like Videos etc.). It's a lot of work to do. But there is no guarantee that Google approves this review. So upfront it is not sure, KeeAnywhere would ever be approved for unrestricted access. This is very frustrating for me as developer. Phil Crocoll had the same problems with Keepass2Android, see e.g. https://github.com/PhilippC/keepass2android/issues?q=is%3Aissue+google+blocked+is%3Aclosed

My initial intention for KeeAnywhere was to develop a plugin to sync to OneDrive (my only used provider). Very quickly other providers have been added and everything was good. But I noticed that cloud is evolving very fast and the amount of work to do raised.

So the question now is:
Is someone out here that would like to take over this approval part - with unknown outcome - and support me with the Google parts in KeeAnywhere?

**Comment by CrashO86** [comment]
> @CrashO86: Вы думаете, что Google отменит блокировку приложения? Я не думаю, что это произойдет. И это не ошибка - скорее решенная проблема. Пожалуйста, переключитесь на GDrive Restricted.

Are you suggesting that I move the database to the root directory of Google Drive? This is a solution to create chaos.

**Comment by Skimmer90** [comment]
Same here.
![image](https://github.com/user-attachments/assets/c802e3e2-9332-4c9a-b20f-0f34a5f9c744)

KP4W: v2.57.1
KA v 2.10 - using GD or GDR.
Browsers: Google Chrome, Firefox, Brave, and Thorium [all current versions as of 20241127]

GD 2FA enabled to use of push notifications to phone.

**Comment by Sepp-PS** [comment]
@Skimmer90 :
you can't open an old GD version via KA, you must open it directly and then you can save it as a new file. Thats all.

**Comment by IzoFox** [comment]
> @meokey Actually I tied it and for me it does work as @ItalianInPoland described with GDrive restricted.
> 
> Steps:
> 
> 1. Add GDrive restricted in KeeAnywhere settings (I used the same google acc't as normal GDrive)
> 2. Save as... -> Save to cloud drive -> GDrive restricted -> The folder you see is empty, just save it there
> 3. You should now see the new file in your GDrive, in the root folder, besides all the other stuff that was there but KeeAnywhere couldn't see
> 4. In Keepass2Android select Open file -> Google Drive
> 5. You should be able to select the new file because Keepass2Android does not run in restricted mode (yet?)
> 6. On Keepass Desktop, you might also have to reopen the file from GDrive restricted.
> 
> That's it. Works for now and syncs both ways.

Thanks, it works for me as a temporarily solution.

**Comment by DrainLife0** [comment]
> @meokey Actually I tied it and for me it does work as @ItalianInPoland described with GDrive restricted.
> 
> Steps:
> 
>     1. Add GDrive restricted in KeeAnywhere settings (I used the same google acc't as normal GDrive)
> 
>     2. Save as... -> Save to cloud drive -> GDrive restricted -> The folder you see is empty, just save it there
> 
>     3. You should now see the new file in your GDrive, in the root folder, besides all the other stuff that was there but KeeAnywhere couldn't see
> 
>     4. In Keepass2Android select Open file -> Google Drive
> 
>     5. You should be able to select the new file because Keepass2Android does not run in restricted mode (yet?)
> 
>     6. On Keepass Desktop, you might also have to reopen the file from GDrive restricted.
> 
> 
> That's it. Works for now and syncs both ways.

4. Doesn't work for me, the app is block, just like in Windows :-(

**Comment by DrainLife0** [comment]
> > @meokey Actually I tied it and for me it does work as @ItalianInPoland described with GDrive restricted.
> > Steps:
> > 
> > 1. Add GDrive restricted in KeeAnywhere settings (I used the same google acc't as normal GDrive)
> > 2. Save as... -> Save to cloud drive -> GDrive restricted -> The folder you see is empty, just save it there
> > 3. You should now see the new file in your GDrive, in the root folder, besides all the other stuff that was there but KeeAnywhere couldn't see
> > 4. In Keepass2Android select Open file -> Google Drive
> > 5. You should be able to select the new file because Keepass2Android does not run in restricted mode (yet?)
> > 6. On Keepass Desktop, you might also have to reopen the file from GDrive restricted.
> > 
> > That's it. Works for now and syncs both ways.
> 
> Thanks, it works for me as a temporarily solution.

I don't understand why it's working for some people... GD only work in restricted mode on KP2android too :-(

**Comment by A-n-d-y-K** [comment]
Hi hi, I´m facing the same issue as @Skimmer90 reported with [comment](https://github.com/Kyrodan/KeeAnywhere/issues/438#issuecomment-2504608391) 3 weeks ago.
I also do have same environment:

- OS and Version: Windows 11
- KeePass-Version: 2.57.1
- KeeAnywhere-Version: 2.1.0

and I´m using Keepass 2 + plugin KeeAnywhere since years.
There are currently similar issues reported by the this community. A possible solution was to download a *.kdbx file from GDrive to the machine, open it and save it withing KeePass to cloud see [comment form Jackabomb](https://github.com/Kyrodan/KeeAnywhere/issues/441#issuecomment-2549766406).

Sadly it doesn´t work for me.

Could one of you help me out, may by giving more info on "Restricted Mode" for KeeAnywhere within my google account. A HowTo would really help. Do I need to delete it from Google Account and reauthenticate?

Thanks in advance

**Comment by Jackabomb** [comment]
Make sure you are using "Google Drive (Restricted)" as the account type. Try removing and re-adding the account in KeeAnywhere.
You may also have to unlink KeeAnywhere in your Google Account from "third-party apps".

Recommended reading for more info on the restricted mode:
https://github.com/Kyrodan/KeeAnywhere/issues/398
https://github.com/Kyrodan/KeeAnywhere/issues/393#issuecomment-2173120783
https://github.com/Kyrodan/KeeAnywhere/issues/393#issuecomment-2213672207

And for the technical side:
https://developers.google.com/identity/protocols/oauth2/scopes#drive
It comes down to the difference between the `drive` scope and the `drive.file` scope.

**Comment by Jackabomb** [comment]
Glad you got it working!
About the subdirectory, it's not you, that's just how it is. Restricted mode is limited to saving only in the drive root.

This is because KeeAnywhere doesn't have a way to _create_ folders. And working in restricted mode means you can't see folders that _already exist_. So the result is no folders at all.

**Comment by A-n-d-y-K** [comment]
@Jackabomb ,
thanks much for fast feedback and confirmation.

Hope there will be a solution or workaround for having *.kdbx file in a subdirectory. Why? Keeps GDrive more organized

**Comment by Rushmed** [comment]
Now for me the gdrive restricted acces is working.
Is there a way to acces the same file with K2A?

**Comment by willistgt** [comment]
I encounter the same problem since weeks now. For me it is no solution to use GDrive Restricted, because I want to access the database from multiple PCs using KeePass and additionally from different Mobile Phone Apps using different Google Accounts. Are there plans to solve this issue or do I have to search for another alternative? I also can't use Google Drive's Sync&Backup Tool (and then open the kdbx files from local disk) on at least one PC.

**Comment by Jackabomb** [comment]
@Kyrodan, I am still interested in working on the app review approval.
If this is acceptable, please let me know what needs to be done, how it has gone so far, etc?
It sounds like they sent you some info/announcement about the change. Maybe sharing/sending me that info would be a good place to start.
Maybe we could make a new issue to coordinate that effort, or find some other communications channel than here to discuss.

**Comment by nascentt** [comment]
>  Have you tried to remove your 3rd party app registrations and create a new one? 

I have. I can't see the restricted keepass.kdbx to select anymore outside of keeanywhere

The previous db shows an unavailable for a storage.legacy path.

I think the restricted path creates by keeanywhere is now only accessible by keeanywhere and other apps can no longer see those files.

**Comment by Sepp-PS** [comment]
I have understood the restricted mode as you described. You must open the file via Keepass and Google Drive. Then you can store it again via the restricted mode and all can be synced via the old solution.
Or with other words: You can brows only the "own files".

**Comment by nascentt** [comment]
again, I've already selected restricted mode in keeanywhere and am successfully syncing there, but i am unable to then access the database form other apps to sync with other keepass instances.

**Comment by Jackabomb** [comment]
@nascentt You understand correctly. We can sync between two devices both using KeeAnywhere, no problem (_if_ KeeAnywhere is registered with Google Drive as the file's creator, which is often not the case initially). We cannot sync between two different apps* right now because KeeAnywhere can only see "its own files", which means we cannot see files created by other apps.

*It is still possible to sync with certain other apps, depending on what permissions level the other app has. If the other app has full (non-restricted) access to your Google Drive, then we can sync with it. If the other app _also_ is restricted just like KeeAnywhere, then the two cannot share files. Most apps have the lower, restricted permissions because they're in the same boat we are. But I imagine a few out there have successfully passed Google's new review process and regained approval for the higher access level.

@Kyrodan I am still interested in helping with Google Drive functionality, including the review process. I still use Google Drive+KeeAnywhere daily.
